src/array.rs

```diff
@@ -106,62 +29,67 @@ const fn digits(x: usize) -> usize {
     num_digits
 }
 
-// This overrides the impl on [T; N] through Deref
-impl<T: Miniconf, const N: usize> Miniconf for Array<T, N> {
-    fn name_to_index(value: &str) -> core::option::Option<usize> {
-        value.parse().ok()
-    }
-
-    fn get_by_key<K, S>(&self, mut keys: K, ser: S) -> Result<usize, Error<S::Error>>
-    where
-        K: Iterator,
-        K::Item: Key,
-        S: serde::Serializer,
-    {
-        let key = keys.next().ok_or(Error::TooShort(0))?;
-        let index = key.find::<Self>().ok_or(Error::NotFound(1))?;
-        let item = self.0.get(index).ok_or(Error::NotFound(1))?;
-        item.get_by_key(keys, ser).increment()
-    }
-
-    fn set_by_key<'a, K, D>(&mut self, mut keys: K, de: D) -> Result<usize, Error<D::Error>>
-    where
-        K: Iterator,
-        K::Item: Key,
-        D: serde::Deserializer<'a>,
-    {
-        let key = keys.next().ok_or(Error::TooShort(0))?;
-        let index = key.find::<Self>().ok_or(Error::NotFound(1))?;
-        let item = self.0.get_mut(index).ok_or(Error::NotFound(1))?;
-        item.set_by_key(keys, de).increment()
-    }
-
-    fn traverse_by_key<K, F, E>(mut keys: K, mut func: F) -> Result<usize, Error<E>>
-    where
-        K: Iterator,
-        K::Item: Key,
-        F: FnMut(usize, &str) -> Result<(), E>,
-    {
-        let key = keys.next().ok_or(Error::TooShort(0))?;
-        let index = key.find::<Self>().ok_or(Error::NotFound(1))?;
-        if index >= N {
-            return Err(Error::NotFound(1));
+macro_rules! depth {
+    ($($d:literal)+) => {$(
+        impl<T: Miniconf<{$d - 1}>, const N: usize> Miniconf<$d> for [T; N] {
+            fn name_to_index(value: &str) -> core::option::Option<usize> {
+                value.parse().ok()
+            }
+
+            fn get_by_key<K, S>(&self, mut keys: K, ser: S) -> Result<usize, Error<S::Error>>
+            where
+                K: Iterator,
+                K::Item: Key,
+                S: serde::Serializer,
+            {
+                let key = keys.next().ok_or(Error::TooShort(0))?;
+                let index = key.find::<$d, Self>().ok_or(Error::NotFound(1))?;
```

Comment:
Flexibility and stability.
E.g. if you come up with a efficient scheme of packing key indices, you can just implement your unpacker to usize in `find` in your own crate 
Otherwise you'd have to change the trait or even touch each existing impl.

Reply:
Flexibility and stability.
E.g. if you come up with a efficient scheme of packing key indices, you can just implement your unpacker to usize in `find` in your own crate 
Otherwise you'd have to change the trait or even touch each existing impl.